### PR TITLE
Added NifiNodeGroupAutoscaler with basic scaling strategies

### PR DESCRIPTION
**Pull request by michael81877**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | 
| License         | Apache 2.0

### What's in this PR?

This PR contains an implementation of the autoscaling design voted on here: https://konpytika.slack.com/archives/C035FHN1MNG/p1650031443231799

It adds:
- A new custom resource called `NifiNodeGroupAutoscaler`
- Adds `Nodes.Spec.Labels` so that nodes in a `NifiCluster` can be tagged with arbitrary labels and specifically for the `NifiNodeGroupAutoscaler` to identify nodes to manage
- A new controller for this custom resource. Its reconciliation loop is as follows:
  - Fetch the current `NifiCluster.Spec.Nodes` list, filter it by the nodes within to manage by the provided `NifiNodeGroupAutoscaler.Spec.NodeLabelsSelector` 
  - Compare the managed node list with the current `NifiNodeGroupAutoscaler.Spec.Replicas` setting and determine if scaling needs to happen. If the # replicas > # managed nodes, then scale up. If the # replicas < # managed nodes, then scale down. Else do nothing.
  - If a scale event happened, patch the `NiFiCluster.Spec.Nodes` list with the added/removed nodes & update the scale subresource status fields. 
- A `NifiNodeGroupAutoscaler` can manage any subset of nodes in a `NifiCluster` up to the entire cluster. With this, you can enable highly resourced (mem/cpu) node groups for volume bursts or to just autoscale entire clusters driven by metrics.
- I significantly reduced the verbosity of the logs by default since it was difficult to track what the operator was actually doing. In general, i think this should be continued to the point where the operator _only_ logs what actions it has taken or changes it has made in k8s.

I don't necessarily consider this PR complete, which is why its in draft status. See the additional context below. However, i have tested this on a live system and it does work.

### Why?

To enable horizontal scaling of nifi clusters.

### Additional context

There are a few scenarios that need to be addressed prior to merging this:

- On a scale up event, when the `NifiNodeGroupAutoscaler` adds nodes to the `NifiCluster.Spec.Nodes` list, the nifi cluster controller correctly adds a new pod to the deployment. However, when that node comes completely up and is `Running` in k8s, nifikop kills it and kicks off a `RollingUpgrade` and basically restarts the new node. This occurs [here](https://github.com/konpyutaika/nifikop/blob/master/pkg/resources/nifi/nifi.go#L677), but i'm not exactly sure what causes it. Scaling down happens "gracefully"

- It's not possible to deploy a `NifiCluster` with only autoscaled node groups. The `NifiCluster` CRD requires that you specify at least one node in the `spec.nodes` list. Do we want to support this? If so, we may need to adjust the cluster initialization logic in the `NifiCluster` controller.

I did all of my testing via ArgoCD. When the live state differs from the state in git, ArgoCD immediately reverts it so I had to craft my applications carefully to avoid ArgoCD undoing the changes that the `HorizontalPodAutoscaler` and `NifiNodeGroupAutoscaler` controllers were making.

I've tested scaling up and down and successfully configured the `HorizontalPodAutoscaler` to pull nifi metrics from Prometheus. Here's the autoscaler config that i used for that setup:

```yaml
apiVersion: nifi.konpyutaika.com/v1alpha1
kind: NifiNodeGroupAutoscaler
metadata:
  labels:
    argocd.argoproj.io/instance: as-nf
  name: as-nf-scale-group
  namespace: nifi
spec:
  clusterRef:
    name: as-nf
    namespace: nifi
  nodeConfigGroupId: scale-group
  nodeLabelsSelector:
    matchLabels:
      scale_me: 'true'
  upscaleStrategy: simple
  downscaleStrategy: lifo
```

And the `nifi_amount_items_queued_sum` metric is added to the prometheus-adapter deployment as follows:

```yaml
prometheus-adapter:
  rules:
    custom:
    - seriesQuery: 'nifi_amount_items_queued'
      resources:
        # skip specifying generic resource<->label mappings, and just
        # attach only pod and namespace resources by mapping label names to group-resources
        overrides:
          kubernetes_namespace:
            resource: "namespace"
          kubernetes_pod_name: 
            resource: "pod"
      name:
        matches: "^(.*)"
        as: "${1}_sum"
      metricsQuery: (sum(<<.Series>>{<<.LabelMatchers>>}) by (<<.GroupBy>>))
```

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)
- [x] Append changelog with changes

### CODE REVIEW

File: pkg/resources/nifi/pod.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ const (
 	ContainerName string = "nifi"
 )
 
-func (r *Reconciler) pod(id int32, nodeConfig *v1alpha1.NodeConfig, pvcs []corev1.PersistentVolumeClaim, log logr.Logger) runtimeClient.Object {
+func (r *Reconciler) pod(node v1alpha1.Node, nodeConfig *v1alpha1.NodeConfig, pvcs []corev1.PersistentVolumeClaim, log logr.Logger) runtimeClient.Object {
```
Comment (erdrix) [inline pkg/resources/nifi/pod.go:43]: Why this change ?

File: api/v1alpha1/nifinodegroupautoscaler_types.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	// reference to the nodeConfigGroup that will be managed and autoscaled
+	// This Id is used to compute the names of some Kubernetes resources, so it must be a safe value.
+	// +kubebuilder:validation:Pattern:="[a-z0-9]([-a-z0-9]*[a-z0-9])?"
+	// +kubebuilder:validation:MaxLength:=63
+	NodeConfigGroupId string `json:"nodeConfigGroupId"`
+	// A label selector used to identify & manage Node objects in the referenced NifiCluster. Any node matching this selector will be managed by this autoscaler. Even if that node was previously statically defined.
+	NodeLabelsSelector *metav1.LabelSelector `json:"nodeLabelsSelector"`
+	// the node readOnlyConfig for each node in the node group
+	// +optional
+	ReadOnlyConfig *ReadOnlyConfig `json:"readOnlyConfig,omitempty"`
```
[the remark this answers is not in the document]
Comment (michael81877) [reply api/v1alpha1/nifinodegroupautoscaler_types.go:40]: What should the behavior be if both a `nodeConfigGroupId` and a `nodeGroup` are configured since they would conflict with each other?

File: pkg/resources/nifi/pod.go (hunk 1 above)
Comment (michael81877) [reply pkg/resources/nifi/pod.go:43]: I passed the `Node` into this method instead of just its ID so that i had access to the node's labels (a new field) on line 123.

File: pkg/util/autoscale/strategy.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+// ScaleDown takes the current set of nodes, removes numNodesToRemove per the strategy, and
+// returns the set of nodes that should be removed from the set of currentNodes
+func (lifo *LIFOHorizontalDownscaleStrategy) ScaleDown(currentNodes []v1alpha1.Node, numNodesToRemove int32) (nodesToRemove []v1alpha1.Node, err error) {
+	numberOfCurrentNodes := int32(len(currentNodes))
+	if numNodesToRemove >= numberOfCurrentNodes || numNodesToRemove == 0 {
+		return []v1alpha1.Node{}, nil
+	}
+
+	nodesToRemove = []v1alpha1.Node{}
+	nodesToRemove = append(nodesToRemove, currentNodes[numberOfCurrentNodes-numNodesToRemove:]...)
```
[the remark this answers is not in the document]
Comment (michael81877) [reply pkg/util/autoscale/strategy.go:36]: Fair point. Do you think it'd be sufficient to add a sortable `creationTimestamp` to the [NodeState](https://github.com/konpyutaika/nifikop/blob/6206ded01f2547d7f8ffe53da80de95fe7e34b9b/api/v1alpha1/common_types.go#L319)? Then we could order nodes which are ready by the time they were created and remove the last one that was added to the cluster. That might be a less complex approach than maintaining a separate ordered list.

Thoughts?

File: controllers/nifinodegroupautoscaler_controller.go [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+		}
+		o, err = patch.DefaultAnnotator.GetOriginalConfiguration(nodeGroupAutoscaler)
+	}
+
+	// Check if the cluster reference changed.
+	original := &v1alpha1.NifiNodeGroupAutoscaler{}
+	json.Unmarshal(o, original)
+	if !v1alpha1.ClusterRefsEquals([]v1alpha1.ClusterReference{original.Spec.ClusterRef, nodeGroupAutoscaler.Spec.ClusterRef}) {
+		nodeGroupAutoscaler.Spec.ClusterRef = original.Spec.ClusterRef
+	}
```
[the remark this answers is not in the document]
Comment (michael81877) [reply controllers/nifinodegroupautoscaler_controller.go:122]: I agree - I will  drop this reference check logic. No need to check if its changed.

File: controllers/nifinodegroupautoscaler_controller.go [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+
+	// Handle HorizontalAutoScaler
+	hpa, err := r.horizontalPodAutoscaler(r.Log, cluster, nodeGroupAutoscaler)
+	if err != nil {
+		return RequeueWithError(r.Log, "failed to generate horizontal pod autoscaler spec", err)
+	}
+	err = k8sutil.Reconcile(r.Log, r.Client, hpa, nil)
+	if err != nil {
+		return RequeueWithError(r.Log, "failed to reconcile horizontal pod autoscaler", err)
+	}
```
[the remark this answers is not in the document]
Comment (michael81877) [reply controllers/nifinodegroupautoscaler_controller.go:145]: I think this is a good point. I was just following what seemed to be the approach for the `NifiCluster` since it creates things like `PodDisruptionBudget`.

I guess if we didn't create the HPA here, then users would have to create a `NifiNodeGroupAutoscaler` and a `HorizontalPodAutoscaler` with a `scaleTargerRef` that's pointed at that `NifiNodeGroupAutoscaler`. 

I suppose for the same reason you may not want to accept PR #84 to have the operator create a `ServiceMonitor` and rather leave that to a helm chart. Is that correct?

File: controllers/nifinodegroupautoscaler_controller.go [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+	numCurrentReplicas := int32(len(currentReplicas))
+
+	// if the current number of nodes being managed by this autoscaler is different than the replica setting,
+	// then set the autoscaler status to out of sync to indicate we're changing the NifiCluster node config
+	if numDesiredReplicas != numCurrentReplicas {
+		r.Log.Info(fmt.Sprintf("Replicas changed from %d to %d", numCurrentReplicas, numDesiredReplicas))
+		if err = r.updateAutoscalerReplicaState(ctx, nodeGroupAutoscaler, v1alpha1.AutoscalerStateOutOfSync); err != nil {
+			return RequeueWithError(r.Log, fmt.Sprintf("Failed to udpate node group autoscaler state for node group %s", nodeGroupAutoscaler.Spec.NodeConfigGroupId), err)
+		}
```
[the remark this answers is not in the document]
Comment (michael81877) [reply controllers/nifinodegroupautoscaler_controller.go:161]: Good catch! I think it would be sufficient to broaden the surrounding if-block to say:

```go
// If the replica setting doesn't match the node group size or if we're currently out of sync, then scale up/down
if numDesiredReplicas != numCurrentReplicas || autoscaler.Status.State == v1alpha1.AutoscalerStateOutOfSync {
  // ..... scale up/down .....
}
```

Do you agree?

File: controllers/nifinodegroupautoscaler_controller.go [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+		return RequeueWithError(r.Log, fmt.Sprintf("Failed to update node group autoscaler replica status for node group %s", nodeGroupAutoscaler.Spec.NodeConfigGroupId), err)
+	}
+
+	return reconcile.Result{
+		RequeueAfter: util.GetRequeueInterval(r.RequeueInterval, r.RequeueOffset),
+	}, nil
+}
+
+// scaleUp updates the provided cluster.Spec.Nodes list with the appropriate numNodesToAdd according to the autoscaler.Spec.UpscaleStrategy
+func (r *NifiNodeGroupAutoscalerReconciler) scaleUp(autoscaler *v1alpha1.NifiNodeGroupAutoscaler, cluster *v1alpha1.NifiCluster, numNodesToAdd int32, startingNodeId int32) error {
```
[the remark this answers is not in the document]
Comment (michael81877) [reply controllers/nifinodegroupautoscaler_controller.go:213]: I agree - Do you have a preference for answering the question "how do we choose new node ids"? Users can choose whatever node Ids they want for the initial nodes (e.g. 1, 2, 50, 7000), so I see a couple of options:

1. We order the node Ids in the node group and iterate until we find a gap, like you suggest in [this](https://github.com/konpyutaika/nifikop/pull/89#discussion_r872537245) comment. 

2. We always choose higher node Ids. Order the node Ids, find the maxId and every new node Id is relative to that maxId. 

With either approach, we'd need to ensure that the node Id isn't being used by another node group.

File: controllers/nifinodegroupautoscaler_controller.go [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+
+func (r *NifiNodeGroupAutoscalerReconciler) removeFinalizer(ctx context.Context, autoscaler *v1alpha1.NifiNodeGroupAutoscaler) error {
+	autoscaler.SetFinalizers(util.StringSliceRemove(autoscaler.GetFinalizers(), autoscalerFinalizer))
+	_, err := r.updateAndFetchLatest(ctx, autoscaler)
+	return err
+}
+
+func (r *NifiNodeGroupAutoscalerReconciler) finalizeNifiNodeGroupAutoscaler(autoscaler *v1alpha1.NifiNodeGroupAutoscaler) error {
+
+	//TODO: remove nodes from NifiCluster that this auto scaler is controlling?
```
[the remark this answers is not in the document]
Comment (michael81877) [reply controllers/nifinodegroupautoscaler_controller.go:375]: I agree. Nothing should happen if the node group autoscaler is deleted. I'll adjust accordingly

File: api/v1alpha1/nifinodegroupautoscaler_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (erdrix) [reply api/v1alpha1/nifinodegroupautoscaler_types.go:40]: They will be merged with a higher priority for the `NodeConfig`.

File: controllers/nifinodegroupautoscaler_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (erdrix) [reply controllers/nifinodegroupautoscaler_controller.go:145]: Yes, exactly! 
(As soon as I find some time to do this, I'll add a helm chart in the repo that wrap the NiFiCluster and ServiceMonitor definition, which is why I haven't responded to your PR yet).

The `PodDisruptionBudget` is handled by the operator, as we directly handle the definition of a node that is a `Pod` and it is closely coupled.

File: controllers/nifinodegroupautoscaler_controller.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (erdrix) [reply controllers/nifinodegroupautoscaler_controller.go:161]: Yup :)

File: controllers/nifinodegroupautoscaler_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (erdrix) [reply controllers/nifinodegroupautoscaler_controller.go:213]: There are pros and cons for both options, as long as we keep the logic centralized in one place, we can make an arbitrary choice without thinking too much and see how it works (it will be easy to change it later!). 

So no preference, you can choose the one you like best :D

File: pkg/util/autoscale/strategy.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (erdrix) [reply pkg/util/autoscale/strategy.go:36]: LGTM

File: api/v1alpha1/nifinodegroupautoscaler_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (michael81877) [reply api/v1alpha1/nifinodegroupautoscaler_types.go:40]: Sounds good!

File: controllers/nifinodegroupautoscaler_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (michael81877) [reply controllers/nifinodegroupautoscaler_controller.go:145]: That makes sense. I will close #84.

File: controllers/nifinodegroupautoscaler_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (michael81877) [reply controllers/nifinodegroupautoscaler_controller.go]: I elected to go with the approach you suggested (i.e. 1.) and implemented that. Let me know your thoughts!

File: controllers/nifinodegroupautoscaler_controller.go [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+		// The optimistic lock here ensures that we only patch the latest version of the NifiCluster to avoid stomping on changes any other process makes.
+		// Ideally, we could use a strategic merge, but it's not supported for CRDs: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility
+		clusterPatch := runtimeClient.MergeFromWithOptions(cluster.DeepCopy(), runtimeClient.MergeFromWithOptimisticLock{})
+
+		if numDesiredReplicas > numCurrentReplicas {
+			// need to increase node group
+			numNodesToAdd := numDesiredReplicas - numCurrentReplicas
+			r.Log.Info(fmt.Sprintf("Adding %d more nodes to cluster %s spec.nodes configuration for node group %s", numNodesToAdd, cluster.Name, nodeGroupAutoscaler.Spec.NodeConfigGroupId))
+
+			startingNodeId := maxId + 1
```
[the remark this answers is not in the document]
Comment (michael81877) [reply controllers/nifinodegroupautoscaler_controller.go]: This is what i implemented 😄

File: api/v1alpha1/nifinodegroupautoscaler_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (michael81877) [reply api/v1alpha1/nifinodegroupautoscaler_types.go:39]: Ah, i didn't realize this was already implemented in `NifiCluster`, so no need to add any additional logic. I will just add `NodeConfig` to the autoscaler spec.

node config merging: https://github.com/konpyutaika/nifikop/blob/master/pkg/util/util.go#L164-L178

usage: https://github.com/konpyutaika/nifikop/blob/master/pkg/resources/nifi/nifi.go#L140

**Comment by michael81877** [comment]
I'm happy to squash all of these commits if that's desired.

**Comment by michael81877** [comment]
Just following up on this PR. I believe it's functionally complete with one exception. 

When a node group is scaled up (e.g. the NifiCluster.spec.nodes list is patched), NiFiKop will correctly create the new pod. However, when the pod status eventually is ready & running, NiFiKop kills the pod and runs a graceful upscale cluster task effectively restarting the new pod. 

Is there a way we can tell the operator to run an upscale operation to begin with to avoid this pod restarting?

**Comment by erdrix** [comment]
> Just following up on this PR. I believe it's functionally complete with one exception.
> 
> When a node group is scaled up (e.g. the `NifiCluster.spec.nodes` list is patched), NiFiKop will correctly create the new pod. However, when the pod status eventually is ready & running, NiFiKop kills the pod and runs a graceful upscale cluster task effectively restarting the new pod.
> 
> Is there a way we can tell the operator to run an upscale operation to begin with to avoid this pod restarting?
> 
> Once this is resolved, i will remove the draft status of this PR so that we can move to merge it.

I will merge this PR on a branch and create a PR to remove the pod restart because with the latest version on NiFi it is no longer necessary.

**Review by erdrix — APPROVED** [review]

**Comment by michael81877** [comment]
> > Just following up on this PR. I believe it's functionally complete with one exception.
> > When a node group is scaled up (e.g. the `NifiCluster.spec.nodes` list is patched), NiFiKop will correctly create the new pod. However, when the pod status eventually is ready & running, NiFiKop kills the pod and runs a graceful upscale cluster task effectively restarting the new pod.
> > Is there a way we can tell the operator to run an upscale operation to begin with to avoid this pod restarting?
> > Once this is resolved, i will remove the draft status of this PR so that we can move to merge it.
> 
> I will merge this PR on a branch and create a PR to remove the pod restart because with the latest version on NiFi it is no longer necessary.

Sounds good! I can help review & test these changes on that PR.
